Undefined subroutine &%s called, close to label '%s'

richardleach opened the pull request:
https://github.com/Perl/perl5/issues/17839 requested a compile-time
warning for the situation whereby a single colon is accdentally typed
as a package separator when calling a function. For example:
```
package BOOP;
sub beep;
package main;
BOOP:beep();  # Meant to be BOOP::beep();
```
However, because of both Perl's syntax and the potential to populate the
stash at runtime, this falls somewhere between very difficult and
impossible. As an alternative, some enhanced fatal error wording was
requested and these commits attempt to provide that.

The above example would previously die with the message:
```
Undefined subroutine &main::beep called at ... line 4.
```
Now it dies with the message:
```
Undefined subroutine &main::beep called, close to label 'BOOP' at ... line 4.
```

For some of the same reasons mentioned, distinguishing this typo from
other errors at runtime - such as the target subroutine not being
present at all - is also nigh on impossible. The hope is that the
error message will give some additional clue when the error is the
result of a typo, without distracting the user in all other cases.

As part of these commits, some common `DIE()` calls in `pp_entersub` were
extracted into a new helper function, `S_croak_undefined_subroutine`, to
avoid adding (and slightly reduce) cold code bloating in `pp_entersub`.

iabyn reviewed (commented):
I'd prefer to see this split into two commits - the first to extract out the DIEes into a separate helper function, and the second to add the LABEL functionality.

I think the perldiag text (and perldelta) could be improved: give an actual example of the problem it's trying to flag, e.g. "C<foo:bar()> would be parsed as a label followed by an unqualified function name: C<foo:  bar()>".

I'm somewhat twitchy that the message is generated even if the label doesn't immediately precede the function call but is merely in the same statement; E.g.

FOO: if ($x > 1 && foo()) { ... }

richardleach commented:
> I'd prefer to see this split into two commits - the first to extract out the DIEes into a separate helper function, and the second to add the LABEL functionality.
>
> I think the perldiag text (and perldelta) could be improved: give an actual example of the problem it's trying to flag, e.g. "Cfoo:bar() would be parsed as a label followed by an unqualified function name: C<foo: bar()>".

I'll make those changes now.

> I'm somewhat twitchy that the message is generated even if the label doesn't immediately precede the function call but is merely in the same statement;

Yes, I'm in two minds about it for that reason too. If, on balance, it looks like this will add more confusion than clarity, there's no point in adding it.

richardleach committed:
Move pp_entersub DIE() code into S_croak_undefined_subroutine

In `pp_entersub`, when trying to find the subroutine `CV` to execute,
there are three nearby code paths that can result in a `DIE()`. This
commit extracts the DIE() logic to a helper subroutine,
`S_croak_undefined_subroutine`. This is partly in anticipation of
additional warning logic, but also generally reduces the size of this
hot function.

richardleach committed:
Undefined subroutine &%s called, close to label '%s'

https://github.com/Perl/perl5/issues/17839 requested a compile-time
warning for the situation whereby a single colon is accdentally typed
as a package separator when calling a function. For example:
```
package BOOP;
sub beep;
package main;
BOOP:beep();  # Meant to be BOOP::beep();
```
However, because of both Perl's syntax and the potential to populate the
stash at runtime, this falls somewhere between very difficult and
impossible. As an alternative, some enhanced fatal error wording was
requested and this commit attempts to provide that.

The above example would previously die with the message:
```
Undefined subroutine &main::beep called at ... line 4.
```
Now it dies with the message:
```
Undefined subroutine &main::beep called, close to label 'BOOP' at ... line 4.
```

For some of the same reasons mentioned, distinguishing this typo from
other errors at runtime - such as the target subroutine not being
present at all - is also nigh on impossible. The hope is that the
error message will give some additional clue when the error is the
result of a typo, without distracting the user in all other cases.

iabyn commented:
On Mon, Dec 16, 2024 at 01:43:30PM -0800, Richard Leach wrote:
> I'll make those changes now.

Unless I'm misreading it, the 'detect label' code is actually part of the
"Move pp_entersub DIE() code into S_croak_undefined_subroutine" commit.

> > I'm somewhat twitchy that the message is generated even if the label doesn't immediately precede the function call but is merely in the same statement;
> 
> Yes, I'm in two minds about it for that reason too. If, on balance, it looks like this will add more confusion than clarity, there's no point in adding it.

Perhaps the warning text should only added for the case where

    PL_curcop->op_next == PL_op

?

-- 
A power surge on the Bridge is rapidly and correctly diagnosed as a faulty
capacitor by the highly-trained and competent engineering staff.
    -- Things That Never Happen in "Star Trek" #9

richardleach commented:
> Unless I'm misreading it, the 'detect label' code is actually part of the "Move pp_entersub DIE() code into S_croak_undefined_subroutine" commit.

Probably I didn't separate the changes properly, will take a look.

> Perhaps the warning text should only added for the case where PL_curcop->op_next == PL_op ?

I'm not following, isn't `PL_curcop->op_next` likely to point to a `pushmark`?

iabyn commented:
On Mon, Jan 06, 2025 at 02:25:37PM -0800, Richard Leach wrote:
> > Perhaps the warning text should only added for the case where PL_curcop->op_next == PL_op ?
> 
> I'm not following, isn't `PL_curcop->op_next` likely to point to a `pushmark`?

Sorry, I meant

    OpSIBLING(PL_curcop) == PL_op

-- 
You live and learn (although usually you just live).

richardleach commented:
> Perhaps the warning text should only added for the case where 
> [snip]  OpSIBLING(PL_curcop) == PL_op

I'm not sure. That would rule out cases where the sub return value is then assigned, or fed into another operation within the same statement. Both of those are surely pretty common.

iabyn commented:
On Mon, Jan 13, 2025 at 02:52:18PM -0800, Richard Leach wrote:
> > Perhaps the warning text should only added for the case where 
> > [snip]  OpSIBLING(PL_curcop) == PL_op
> 
> I'm not sure. That would rule out cases where the sub return value is
> then assigned, or fed into another operation within the same statement.
> Both of those are surely pretty common.

A COP, and in particular a COP with a label, should only occur at the
start of a statement. We a concerned specifically with a fn call
immediately following a label, .i.e.

    Foo:bar()

If the fn call isn't in void context, then the ambiguity doesn't arise:

    Foo:$x=bar()

If its part of an expression, then unless that expression is in void
context, then against this isn't a problem. And expressions in non-void
context are unusual:

    Foo:bar()+$x

The only sort f expression I can think of which are commonly void are
mutators, but that's not a problem:

    Foo:bar()+=$x

Since this is 'you might not have meant to do this' warning, I'd prefer
that it occasionally skips a warning compared with warning unnecessarily.

I'm happy to be proven wrong, if you can show an uncontrived example of

    Foo:bar()....

(for some ....) which wouldn't be caught by OpSIBLING(PL_curcop) == PL_op



-- 
1 - number of times I have needed to use the cap lock key
10000 - number of times I have CCIDENTLY HIT THE CAPS KEY WHEN TYPING 'A'

richardleach commented:
> A COP, and in particular a COP with a label, should only occur at the start of a statement. We a concerned specifically with a fn call immediately following a label, .i.e. Foo:bar()

Sorry, yes, you're absolutly right. Not sure what I was thinking. I'll make that change at the next opportunity.

richardleach commented:
> I'll make that change at the next opportunity.

Done.
